JT/12_data_structures/non-lineat_types_exercises/non_linear_types.adoc

```diff
@@ -0,0 +1,288 @@
+}
+
+//After adding all the vertices and edges, the graph can be created completely
+Graph createGraph() {
+    Graph graph = new Graph();
+    graph.addVertex("Bob");
+    graph.addVertex("Alice");
+    graph.addVertex("Mark");
+    graph.addEdge("Bob", "Alice");
+    graph.addEdge("Alice", "Mark");
+    return graph;
+}
+----
+
+.*How to add Elements into the Graph?*
+----
+// A method to add a vertice
+void addVertex(String label) {
+    adjVertices.putIfAbsent(new Vertex(label), new ArrayList<>());
+}
+
+// A method to add an edge
+void addEdge(String label1, String label2) {
+    Vertex v1 = new Vertex(label1);
+    Vertex v2 = new Vertex(label2);
+    adjVertices.get(v1).add(v2);
+    adjVertices.get(v2).add(v1);
+}
+----
+
+.*How to Remove Elements from the Graph?*
+----
+// A method to remove a vertice
+void addVertex(String label) {
+    adjVertices.putIfAbsent(new Vertex(label), new ArrayList<>());
+}
+
+// A remove to add an edge
+void addEdge(String label1, String label2) {
+    Vertex v1 = new Vertex(label1);
+    Vertex v2 = new Vertex(label2);
+    adjVertices.get(v1).add(v2);
+    adjVertices.get(v2).add(v1);
+}
+----
+
+.*How to Traverse Elements the Graph?*
+To traverse a Graph means to start in one vertex, and go along the edges to visit other vertices until all vertices, or as many as possible, have been visited. +
+The two most common ways a Graph can be traversed are:
+
+. Depth First Search (DFS)
+. Breadth First Search (BFS)
+
+DFS is usually implemented using a Stack or by the use of recursion (which utilizes the call stack), while BFS is usually implemented using a Queue.
+More details on these algorithms we covered in the section about trees.
+
+=== Useful links
+
+[NOTE]
+====
+* Some exercises you can find link:Graph.java[here]
+* https://www.baeldung.com/cs/graphs[Graph Data Structures]
+* Learn more about https://www.javatpoint.com/java-graph/[graph types] and https://www.baeldung.com/java-graphs[implementation]
+====
+
+== Hash Table
+
+image::../resources/hash_table.png[width=700,align="center"]
+
+A *hash table* or hash map stores key-value pairs in an array. The key is a unique identifier to access or retrieve
+the value, which is the associated data or information.
+
+.*How to Create the Hash Table?*
+----
+Hashtable<Integer, String> hash1 = new Hashtable<Integer, String>();
+----
+
+.*How to add Elements into the Hash Table?*
+----
+hash1.put(1, "Free");
+hash1.put(2, "Courses");
+hash1.put(3, "on");
+----
+
+.*How to Remove Elements from the Hash Table?*
+----
+hashtable.remove(4);
+----
+
+.*How to Get Elements from the Hash Table?*
+This example shows how to use the `getOrDefault()` function which returns the value to which a specified key is mapped, or a defaultValue if the hashtable contains no mapping for that key.
+
+It takes two parameters: a key and a default value to be printed as a message if the key is not present in the hashtable.
+
+----
+hashtable.getOrDefault(2, "Not Found in the Hashtable")
```

Comment:
Do we need a semicolon here?

Reply:
Yes, because '----' also highlights the block of code

Reply:
Added, thanks